spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java

```diff
@@ -20,7 +20,7 @@
     final Level level;
 
     Result(@NonNull String ruleId, int ruleIndex, Message message, List<Location> locations, @NonNull Level level) {
-        this.ruleId = Objects.requireNonNull(ruleId);
+        this.ruleId = String.format("SPOTBUGS_%d", Objects.requireNonNull(ruleId).hashCode());
```

Comment:
Not sure the `rule id` can contain underscore or not 🤔 .
The [official doc](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#def_rule_id) says nothing about acceptable characters.

Reply:
Note for test: hashcode could be negative.

Reply:
The rule id can have underscore, e.g. "EC_UNRELATED_TYPES" is the rule id without the change in this PR.
The negative hash code is valid hash code, if change it e.g. math.abs may increase the possibility of a collision. 
cc @michaelcfanning @eddynaka what do you think about the opaque rule id?

Reply:
These ids need work. The prefix is too long. We shouldn't have underscores or hyphens in the name. Ideally, an opaque id starts with a well-known initial value, 1001 is typical, and then monotonically increases as each rule is added. So, something like this SB1001, SB1002, SB1003, would work best.

A useful convention is to retain rule ids of 0001 - 0999 for internal tool/engine errors.

I can appreciate that my idea requires a more intrusive change, we may need to consider creating a formal place for the rule id in the object model and to populate SARIF from this.

Reply:
In this case we probably need to assign the fixed Ids in meta data in SpotBugs and use these Ids as rule id, and the id always points to same rule.
E.g. in findbugs.xml
<BugPattern **ruleId="SB1001"** abbrev="SBSC" type="SBSC_USE_STRINGBUFFER_CONCATENATION" category="PERFORMANCE"/>
<BugPattern **ruleId="SB1002"**  abbrev="IIL" type="IIL_ELEMENTS_GET_LENGTH_IN_LOOP" category="PERFORMANCE" experimental="true"/>
...
Let me know what do you think?